File: AlgoSolutions/task-assignment.kt

//https://www.algoexpert.io/questions/task-assignment
package com.algoexpert.program


//time - O(nlogn)
//space - O(n)
fun taskAssignment(k: Int, tasks: List<Int>): List<List<Int>> {
    //sort tasks asc duration O(nlogn)
    val sortedTasks = tasks.withIndex().sortedBy { it.value }
    //pairs lowest and highest duration
    var left = 0
    var right = sortedTasks.size - 1
    val pairs = mutableListOf<List<Int>>()
    //O(n)
    while (left < right) {
        pairs.add(listOf(sortedTasks[left].index, sortedTasks[right].index))
        left++
        right--
    }

    return pairs
}


// fun testTaskAssignment() {
//    val tasks = listOf<Int>(1, 3, 5, 3, 1, 4)
//    val result =  taskAssignment(3, tasks);
//     println(result)
// }

File: AlgoSolutions/task-assignment.swift
//https://www.algoexpert.io/questions/task-assignment
class Program {
func taskAssignment(_ k: Int, _ tasks: [Int]) -> [[Int]] {
    //sort tasks map with index and sorted by value
    let sortedTasks = tasks.enumerated()
        .map{($0.offset, $0.element)}
        .sorted(by: {$0.1 < $1.1})
    
    //pairs
    var paris = [[Int]]()
    let count = sortedTasks.count - 1
    
    //list tasks
    for i in 0..<k {
        paris.append([
                      sortedTasks[i].0,
                      sortedTasks[count - i].0
                     ])
    }
 
    return paris
  }


func testTaskAssignment() {
   let tasks = [1, 3, 5, 3, 1, 4]
   let result =  taskAssignment(3, tasks);
    print(result)
}


}
